bindings/src/main/java/io/pravega/storage/gcp/GCPChunkStorage.java

```diff
@@ -0,0 +1,243 @@
+import io.pravega.segmentstore.storage.chunklayer.ChunkNotFoundException;
+import io.pravega.segmentstore.storage.chunklayer.ChunkStorage;
+import io.pravega.segmentstore.storage.chunklayer.ChunkStorageException;
+import io.pravega.segmentstore.storage.chunklayer.ConcatArgument;
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.concurrent.Executor;
+
+/**
+ * {@link ChunkStorage} for GCP based storage.
+ * <p>
+ * Each chunk is represented as a single Object on the underlying storage.
+ * <p>
+ * This implementation works under the assumption that object is created only once and never modified.
+ */
+@Slf4j
+public class GCPChunkStorage extends BaseChunkStorage {
+
+    /**
+     * Error code for not found in google cloud storage
+     */
+    private static final int FILE_NOT_FOUND = 404;
+
+    //region members
+    /**
+     * GCPStorageConfig contains config values.
+     */
+    private final GCPStorageConfig config;
+
+    /**
+     * Storage is provided by GCP to perform cloud operation on google cloud storage.
+     */
+    private final Storage storage;
+
+    //endregion
+
+    //region constructor
+    public GCPChunkStorage(Storage storage, GCPStorageConfig config, Executor executor, boolean shouldCloseClient) {
+        super(executor);
+        this.config = Preconditions.checkNotNull(config, "config");
+        this.storage = Preconditions.checkNotNull(storage, "client");
+    }
+    //endregion
+
+    //region capabilities
+
+    @Override
+    public boolean supportsConcat() {
+        return false;
+    }
+
+    @Override
+    public boolean supportsAppend() {
+        return false;
+    }
+
+    @Override
+    public boolean supportsTruncation() {
+        return false;
+    }
+
+    //endregion
+
+    //region implementation
+
+    @Override
+    protected ChunkHandle doOpenRead(String chunkName) throws ChunkStorageException {
+        if (!checkExists(chunkName)) {
+            throw new ChunkNotFoundException(chunkName, "doOpenRead");
+        }
+        return ChunkHandle.readHandle(chunkName);
+    }
+
+    @Override
+    protected ChunkHandle doOpenWrite(String chunkName) throws ChunkStorageException {
+        if (!checkExists(chunkName)) {
+            throw new ChunkNotFoundException(chunkName, "doOpenWrite");
+        }
+        return new ChunkHandle(chunkName, false);
+    }
+
+    @Override
+    protected int doRead(ChunkHandle handle, long fromOffset, int length, byte[] buffer, int bufferOffset) throws ChunkStorageException {
+        try (ReadChannel readChannel = this.storage.reader(this.config.getBucket(), getObjectPath(handle.getChunkName()))) {
+            try {
+                readChannel.seek(fromOffset);
+                readChannel.limit(fromOffset + length);
+                int val = readChannel.read(ByteBuffer.wrap(buffer, bufferOffset, length));
+                if (val == -1) {
```

Comment:
What does it mean if `val == -1`? It has not read any data? or just a partial read? Add some comments please.